### DISCUSSION TITLE
"pdm sync --clean" does not remove the un-required packages according to the current candidates

### DISCUSSION THREAD

**Discussion by linw1995** [pdm-project/pdm#398] [Q&A] [answered]
I recently use pdm to manage different groups of packages, and use the `pdm sync --clean` command to clean the un-required packages. So I can test my code in different and **isolated** groups of packages. After executing this command the un-required packages are not removed.

The first thing came out of my mind was there maybe have a bug lying on the synchronizer module. But I find this and it is so surprised that this behavior is as design.

>  Remove package only if it is not required by any section 
>  Packages for packaging will never be removed

I see that above comments coming from the below code fragment tell which packages will be to be removed. 

https://github.com/pdm-project/pdm/blob/a3536105a6aae92515130bca5a303e2871464fe2/pdm/installers/synchronizers.py#L113-L126

And my question is why the `pdm sync --clean` does not remove the un-required packages according to the current candidates instead of according to the all candidates?

**Comment by frostming** [comment, marked as answer]
Yes, it is somewhat a hidden feature :P. Because it is a bit annoying to keep in mind which sections are added to installation when doing `pdm sync --clean`, especially for `pdm remove`(with `pdm sync --clean` under the hood) where there is no way to specify which sections to keep.

Say you installed your project with `extra1` feature, and after some development, you realize `foo` is not required anymore. You do this by
`pdm remove foo` and found extra1 is cleaned. This is why I put the comment like that. Also I would like to know what is your case, since `pdm sync` and `pdm remove` are only useful in local development, namely, you are not likely to remove packages in CI or deployment.

An option flag should be possible to address this problem, but I have not come up with an appropriate name for it :P. Any suggestion is welcomed.

**Reply by linw1995** [reply]
IMHO, PDM should give back the control about removing or not for users to use. By adding a new flag `--clean-strict` to remove un-required packages according to the current package candidates. :D

And another thing is that there are `--clean` and `--no-clean` optional flags, so which one is the default behavior?

**Reply by frostming** [reply]
`--no-clean` is the default. This is a useless flag though

**Reply by gdiscry** [reply]
> Also I would like to know what is your case, since `pdm sync` and `pdm remove` are only useful in local development, namely, you are not likely to remove packages in CI or deployment.

In my case, I use `pdm sync --clean --group=…` with Nox to setup the packages installed inside each session. With Nox, it is possible to reuse an existing virtualenv instead of recreating it from scratch, and it saves time. But because `--clean` does not remove the packages still present in the lockfile, there is a risk as the reused virtualenv might be different from a virtualenv created from scratch.

> An option flag should be possible to address this problem, but I have not come up with an appropriate name for it :P. Any suggestion is welcomed.

My suggestions for an option would be something like `--exact`, `--only` or `--strict`.

**Reply by Hnasar** [reply]
I though `--only-keep` made sense too -- check out the change here:
https://github.com/pdm-project/pdm/pull/1191
